src/bin/tools/downloadAndUnzip.ts

```diff
@@ -1,80 +1,289 @@
-import { execSync } from "child_process";
-import * as fs from "fs";
+import { dirname as pathDirname, basename as pathBasename, join as pathJoin } from "path";
+import { createReadStream, createWriteStream, unlinkSync } from "fs";
+import { stat, mkdir, unlink, readFile, writeFile } from "fs/promises";
 import { transformCodebase } from "./transformCodebase";
-import * as crypto from "crypto";
+import { createHash } from "crypto";
+import http from "http";
+import https from "https";
+import { createInflateRaw } from "zlib";
 
-/** assert url ends with .zip */
-export function downloadAndUnzip(params: {
-    isSilent: boolean;
-    url: string;
-    destDirPath: string;
-    pathOfDirToExtractInArchive?: string;
-    cacheDirPath: string;
-}) {
-    const { url, destDirPath, pathOfDirToExtractInArchive, cacheDirPath } = params;
-
-    const extractDirPath = pathJoin(
-        cacheDirPath,
-        `_${crypto.createHash("sha256").update(JSON.stringify({ url, pathOfDirToExtractInArchive })).digest("hex").substring(0, 15)}`
-    );
+import type { Readable } from "stream";
 
-    fs.mkdirSync(cacheDirPath, { "recursive": true });
+function hash(s: string) {
+    return createHash("sha256").update(s).digest("hex");
+}
 
-    const { readIsSuccessByExtractDirPath, writeIsSuccessByExtractDirPath } = (() => {
-        const filePath = pathJoin(cacheDirPath, "isSuccessByExtractDirPath.json");
+async function maybeReadFile(path: string) {
+    try {
+        return await readFile(path, "utf-8");
+    } catch (error) {
+        if ((error as Error & { code: string }).code === "ENOENT") return undefined;
+        throw error;
+    }
+}
 
-        type IsSuccessByExtractDirPath = Record<string, boolean | undefined>;
+async function maybeStat(path: string) {
+    try {
+        return await stat(path);
+    } catch (error) {
+        if ((error as Error & { code: string }).code === "ENOENT") return undefined;
+        throw error;
+    }
+}
 
-        function readIsSuccessByExtractDirPath(): IsSuccessByExtractDirPath {
-            if (!fs.existsSync(filePath)) {
-                return {};
-            }
+/**
+ * Download a file from `url` to `dir`. Will try to avoid downloading existing
+ * files by using an `{hash(url)}.etag` file. If this file exists, we add an
+ * etag headear, so server can tell us if file changed and we should re-download
+ * or if our file is up-to-date.
+ *
+ * Warning, this method assumes that the target filename can be extracted from
+ * url, content-disposition headers are ignored.
+ *
+ * If the target directory does not exist, it will be created.
+ *
+ * If the target file exists and is out of date, it will be overwritten.
+ * If the target file exists and there is no etag file, the target file will
+ * be overwritten.
+ *
+ * @param url download url
+ * @param dir target directory
+ * @returns promise for the full path of the downloaded file
+ */
+async function download(url: string, dir: string): Promise<string> {
```

Comment:
It _might_ seem unorthodox to implement this from scratch, but really, introducing another library was not worth it IMHO

Reply:
Agreed, mutch apreciated!